libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -0,0 +1,801 @@
+AP_GROUPEND
+};
+
+static uavcan::Subscriber<uavcan::equipment::gnss::Fix> *gnss_fix;
+static void gnss_fix_cb(const uavcan::ReceivedDataStructure<uavcan::equipment::gnss::Fix>& msg)
+{
+    if (hal.can_mgr != nullptr) {
+        AP_UAVCAN *ap_uavcan = hal.can_mgr->get_UAVCAN();
+        if (ap_uavcan != nullptr) {
+            AP_GPS::GPS_State *state = ap_uavcan->find_gps_node(msg.getSrcNodeID().get());
+
+            if (state != nullptr) {
+                bool process = false;
+
+                if (msg.status == uavcan::equipment::gnss::Fix::STATUS_NO_FIX) {
+                    state->status = AP_GPS::GPS_Status::NO_FIX;
+                } else {
+                    if (msg.status == uavcan::equipment::gnss::Fix::STATUS_TIME_ONLY) {
+                        state->status = AP_GPS::GPS_Status::NO_FIX;
+                    } else if (msg.status == uavcan::equipment::gnss::Fix::STATUS_2D_FIX) {
+                        state->status = AP_GPS::GPS_Status::GPS_OK_FIX_2D;
+                        process = true;
+                    } else if (msg.status == uavcan::equipment::gnss::Fix::STATUS_3D_FIX) {
+                        state->status = AP_GPS::GPS_Status::GPS_OK_FIX_3D;
+                        process = true;
+                    }
+
+                    const uint64_t posix_offset = UNIX_OFFSET_MSEC;
+                    const uint64_t ms_per_week = 7ULL * 24 * 3600 * 1000;
+                    uint64_t epoch_ms = uavcan::UtcTime(msg.gnss_timestamp).toUSec();
+                    epoch_ms /= 1000;
+                    uint64_t gps_ms = epoch_ms - posix_offset;
+                    state->time_week = (uint16_t) (gps_ms / ms_per_week);
+                    state->time_week_ms = (uint32_t) (gps_ms - (state->time_week) * ms_per_week);
+                }
+
+                if (process) {
+                    Location loc = { };
+                    loc.lat = msg.latitude_deg_1e8 / 10;
+                    loc.lng = msg.longitude_deg_1e8 / 10;
+                    loc.alt = msg.height_msl_mm / 10;
+                    state->location = loc;
+                    state->location.options = 0;
+
+                    state->hdop = msg.pdop;
+
+                    if (!uavcan::isNaN(msg.ned_velocity[0])) {
+                        Vector3f vel(msg.ned_velocity[0], msg.ned_velocity[1], msg.ned_velocity[2]);
+                        state->velocity = vel;
+                        state->ground_speed = norm(vel.x, vel.y);
+                        state->ground_course = wrap_360(degrees(atan2f(vel.y, vel.x)));
+                        state->have_vertical_velocity = true;
+                    } else {
+                        state->have_vertical_velocity = false;
+                    }
+
+                    float pos_cov[9];
+                    msg.position_covariance.unpackSquareMatrix(pos_cov);
+                    if (!uavcan::isNaN(pos_cov[8])) {
+                        state->vertical_accuracy = (pos_cov[8] > 0) ? sqrtf(pos_cov[8]) : -1.0F;
```

Comment:
I see you have changed this. I would have prefered just one _if_ to avoid code duplication, but that's fine.